Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -34,309 +34,269 @@
+bool CurlMultipartHandle::shouldTerminateIfNeeded()
 {
-    auto contentLength = m_buffer.size();
+    if (m_didCompleteMessage) {
+        m_state = State::Terminating;
+        return true;
+    }
 
-    boundaryStartPosition = notFound;
-    lastPartialMatchPosition = contentLength;
+    return false;
+}
 
-    if (contentLength < m_boundary.length())
-        return false;
+CurlMultipartHandle::FindBoundaryResult CurlMultipartHandle::findBoundary()
+{
+    FindBoundaryResult result;
 
-    const auto content = m_buffer.data();
+    auto contentLength = m_buffer.size();
+    const auto contentStartPtr = m_buffer.data();
+    const auto contentEndPtr = contentStartPtr + contentLength;
+
+    auto boundaryLength = m_boundary.length();
+    auto boundaryStartPtr = m_boundary.latin1().data();
+
+    auto matchedStartPtr = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, boundaryStartPtr, boundaryLength));
+    const uint8_t* matchedEndPtr = nullptr;
+    bool needMoreData = false;
+
+    if (matchedStartPtr)
+        matchedEndPtr = matchedStartPtr + boundaryLength;
+    else {
+        for (auto cmpLength = std::min(static_cast<size_t>(boundaryLength - 1), contentLength); cmpLength; --cmpLength) {
+            if (!memcmp(contentEndPtr - cmpLength, boundaryStartPtr, cmpLength)) {
+                needMoreData = true;
+                matchedStartPtr = contentEndPtr - cmpLength;
+                matchedEndPtr = contentEndPtr;
+                break;
+            }
+        }
 
-    for (size_t i = 0; i < contentLength - m_boundary.length(); ++i) {
-        auto length = matchedLength(content + i);
-        if (length == m_boundary.length()) {
-            boundaryStartPosition = i;
-            return true;
+        if (!matchedStartPtr) {
+            // Not found the boundary in the buffer.
+            result.processed = contentLength;
+            return result;
         }
+    }
 
-        if (length)
-            lastPartialMatchPosition = i;
+    // The initial CRLF is considered to be attached to the boundary delimiter line rather than
+    // part of the preceding part. (See RFC2046 [5.1.1. Common Syntax])
+    if (matchedStartPtr - contentStartPtr >= 2 && !memcmp(matchedStartPtr - 2, "\r\n", 2))
+        matchedStartPtr -= 2;
+    else if (matchedStartPtr - contentStartPtr >= 1 && !memcmp(matchedStartPtr - 1, "\n", 1))
+        matchedStartPtr--;
+
+    if (needMoreData || (!m_didCompleteMessage && contentEndPtr - matchedEndPtr < 2)) {
+        // Not enough data to check the boundary or the Close Delimiter.
+        result.processed = matchedStartPtr - contentStartPtr;
+        return result;
     }
 
-    return false;
-}
+    if (contentEndPtr - matchedEndPtr >= 2 && !memcmp(matchedEndPtr, "--", 2)) {
+        result.hasBoundary = true;
+        result.hasCloseDelimiter = true;
+        result.processed = matchedStartPtr - contentStartPtr;
+        result.dataEnd = matchedEndPtr + 2 - contentStartPtr;
```

Comment:
I meant `processed` is the processed length, `dataEnd` is the body data length.
> result.processed = matchedEndPtr + 2 - contentStartPtr;
> result.dataEnd = matchedStartPtr - contentStartPtr;

Reply:
`processed` is confusing. How about `boundaryEnd`?

Reply:
`processed` is the processed length. You can remove the processed length of the buffer. It's not the end of a boundary. 

> m_buffer.remove(0, result.processed);

`dataEnd` is the length of body data. You can send it to the client.

> m_client.didReceiveDataFromMultipart(SharedBuffer::create(m_buffer.data(), result.dataEnd));

Reply:
Fixed